File: solutions/python/relative-distance/1/relative_distance.py
from collections import deque

class RelativeDistance:
    def __init__(self, family_tree):
        self.family_tree = family_tree

    def degree_of_separation(self, person_a, person_b):
        neighbors = {}

        for parent, children in self.family_tree.items():
            if parent not in neighbors:
                neighbors[parent] = set()

            for child in children:
                if child not in neighbors:
                    neighbors[child] = set()

                neighbors[parent].add(child)
                neighbors[child].add(parent)

            for child_a in children:
                for child_b in children:
                    if child_a != child_b:
                        neighbors[child_a].add(child_b)
                        neighbors[child_b].add(child_a)

        if person_a not in neighbors:
            raise ValueError("Person A not in family tree.")

        if person_b not in neighbors:
            raise ValueError("Person B not in family tree.")

        queue = deque([[person_a, 0]])
        visited = {person_a}

        while queue:
            current, degree = queue.popleft()

            if current == person_b:
                return degree

            for neighbor in neighbors[current]:
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append([neighbor, degree + 1])

        raise ValueError("No connection between person A and person B.")

File: solutions/java/relative-distance/1/src/main/java/RelativeDistance.java
import java.util.*;

class RelativeDistance {

    private final Map<String, Set<String>> graph = new HashMap<>();

    public RelativeDistance(Map<String, List<String>> familyTree) {
        for (Map.Entry<String, List<String>> entry : familyTree.entrySet()) {
            String parent = entry.getKey();
            List<String> children = entry.getValue();

            graph.putIfAbsent(parent, new HashSet<>());

            for (String child : children) {
                graph.putIfAbsent(child, new HashSet<>());

                graph.get(parent).add(child);
                graph.get(child).add(parent);
            }


            for (int i = 0; i < children.size(); i++) {
                for (int j = i + 1; j < children.size(); j++) {
                    String childA = children.get(i);
                    String childB = children.get(j);

                    graph.get(childA).add(childB);
                    graph.get(childB).add(childA);
                }
            }
        }
    }

    public int degreeOfSeparation(String personA, String personB) {
        if (!graph.containsKey(personA) || !graph.containsKey(personB)) {
            return -1;
        }

        Queue<PersonLevel> queue = new LinkedList<>();
        Set<String> visited = new HashSet<>();

        queue.add(new PersonLevel(personA, 0));
        visited.add(personA);

        while (!queue.isEmpty()) {
            PersonLevel current = queue.poll();

            if (current.name.equals(personB)) {
                return current.level;
            }

            for (String neighbor : graph.get(current.name)) {
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.add(new PersonLevel(neighbor, current.level + 1));
                }
            }
        }

        return -1;
    }

    private static class PersonLevel {
        String name;
        int level;

        PersonLevel(String name, int level) {
            this.name = name;
            this.level = level;
        }
    }
}

